File: core/causality/causal_recovery_engine.py
from __future__ import annotations

from typing import Any, Dict, List


def recover_causal_runtime(
    causality: Dict[str, Any],
    events: List[Dict[str, Any]],
) -> Dict[str, Any]:
    propagation_order = list(causality.get("propagation_order", []))
    recovered_events = list(events)

    if not propagation_order and events:
        propagation_order = [str(event.get("id", "")) for event in events]

    gaps = []
    for index in range(1, len(propagation_order)):
        if not propagation_order[index]:
            gaps.append(index)

    for gap_index in gaps:
        recovered_events.append({
            "id": f"recovered:evt:{gap_index}",
            "runtime": "recovery",
            "type": "restore",
            "step": gap_index,
            "recovered": True,
        })

    recovered_order = sorted(
        recovered_events,
        key=lambda item: int(item.get("step", 0)),
    )

    return {
        "recovered_events": recovered_order,
        "propagation_order": [
            str(event.get("id", "")) for event in recovered_order
        ],
        "broken_chains_fixed": len(gaps),
        "synchronization_restored": True,
        "bounded": True,
    }

File: src/causality/causalRecoveryEngine.ts
/**
 * Converted from Python: core/causality/causal_recovery_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";

export function recoverCausalRuntime(causality: any, events: any): any {
  var propagation_order: any = [...py.iter(py.get(causality, "propagation_order", []))];
  var recovered_events: any = [...py.iter(events)];
  if ((!py.truthy(propagation_order) && py.truthy(events))) {
    propagation_order = py.iter(events).map((event: any) => py.toStr(py.get(event, "id", "")));
  }
  var gaps: any[] = [];
  var index: any;
  for (index = 1; index < py.len(propagation_order); index++) {
    if (!py.truthy(py.at(propagation_order, index))) {
      py.listAppend(gaps, index);
    }
  }
  var gap_index: any;
  for (gap_index of py.iter(gaps)) {
    py.listAppend(recovered_events, {"id": `recovered:evt:${py.toStr(gap_index)}`, "runtime": "recovery", "type": "restore", "step": gap_index, "recovered": true});
  }
  var recovered_order: any = py.sorted(recovered_events, {key: ((item: any) => py.toInt(py.get(item, "step", 0))) as (item: any) => any});
  return {"recovered_events": recovered_order, "propagation_order": py.iter(recovered_order).map((event: any) => py.toStr(py.get(event, "id", ""))), "broken_chains_fixed": py.len(gaps), "synchronization_restored": true, "bounded": true};
}
